lib/mlir/Dialect/Clift/IR/CliftOps.cpp

```diff
@@ -307,6 +307,70 @@ void SwitchOp::build(OpBuilder &OdsBuilder,
   build(OdsBuilder, OdsState, SignedCaseValues, CaseValues.size());
 }
 
+mlir::ParseResult SwitchOp::parse(OpAsmParser &Parser, OperationState &Result) {
+  // Condition region:
+  Result.addRegion(std::make_unique<Region>());
+
+  // Default case region:
+  Result.addRegion(std::make_unique<Region>());
+
+  if (Parser.parseRegion(*Result.regions[0]).failed())
+    return Parser.emitError(Parser.getCurrentLocation(),
+                            "Expected switch condition region");
+
+  llvm::SmallVector<int64_t, 16> CaseValues;
+  while (Parser.parseOptionalKeyword("case").succeeded()) {
+    uint64_t CaseValue;
+    if (Parser.parseInteger(CaseValue).failed())
+      return Parser.emitError(Parser.getCurrentLocation(),
+                              "Expected switch case value");
+
+    auto R = std::make_unique<Region>();
+    if (Parser.parseRegion(*R).failed())
+      return Parser.emitError(Parser.getCurrentLocation(),
+                              "Expected switch case region");
+
+    CaseValues.push_back(static_cast<uint64_t>(CaseValue));
+    Result.addRegion(std::move(R));
+  }
+
+  if (Parser.parseOptionalKeyword("default").succeeded()) {
+    if (Parser.parseRegion(*Result.regions[1]).failed())
+      return Parser.emitError(Parser.getCurrentLocation(),
+                              "Expected switch default region");
+  }
+
+  Result.attributes.set("case_values",
+                        DenseI64ArrayAttr::get(Parser.getContext(),
+                                               CaseValues));
+
+  if (Parser.parseOptionalAttrDict(Result.attributes).failed())
+    return mlir::failure();
+
+  return mlir::success();
+}
+
+void SwitchOp::print(OpAsmPrinter &Printer) {
+  Printer << ' ';
+  Printer.printRegion(getConditionRegion());
+
+  for (unsigned I = 0, C = getNumCases(); I < C; ++I) {
+    Printer << " case " << getCaseValue(I) << ' ';
+    Printer.printRegion(getCaseRegion(I));
+  }
+
+  if (hasDefaultCase()) {
+    Printer << " default ";
+    Printer.printRegion(getDefaultCaseRegion());
+  }
+
+  static constexpr llvm::StringRef Elided[] = {
+    "case_values",
+  };
+
+  Printer.printOptionalAttrDict(getOperation()->getAttrs(), Elided);
```

Comment:
`switch-with-mismatched-cases.mlir` uses the universal MLIR format to specify `case_values` explicitly. That's required for testing mismatched case values and regions, because using the custom assembly format it is impossible to define such an operation.

Reply:
What's the universal MLIR format

Even without knowing what that is, we're implementing custom serialization/deserialization for `switch`, so I don't think we have any obligation to follow its rules.

Honestly, I'd be much happier if our parser didn't recognize a deserialization format that a) is ugly and unintuitive to read and b) allows parsing stuff that can not verify.
I know that there are probably other things that we can parse and do not verify, but it seems to me that disallowing this format would still allow to express with the other text format (the readable one) any valid `switch` construct that we want to allow.
If so, let's just disable the universal format if possible.

And for what concern the tests, I know that without the universal MLIR format we wouldn't be able to test mismatched cases in `.mlir` files anymore, but that just means that the proper way to test this is still C++ and not `.mlir` files + `FileCheck`.

Reply:
> What's the universal MLIR format
The language reference describes the MLIR notation. See https://mlir.llvm.org/docs/LangRef/#operations `generic-operation`. It's generic notation that can describe any operation, regardless of any custom assembly format.

> Honestly, I'd be much happier if our parser didn't recognize [the generic format]
I'm not sure there's anything we can do to disable it, short of modifying LLVM to remove it entirely. Neither am I convinced that it would be a good idea, even if we could easily do it. This is very much an escape hatch where the user is explicitly stepping outside of the guardrails.

> that just means that the proper way to test this is still C++
I don't agree. In this case the argument about pinning the custom assembly format doesn't even apply. Writing the same test in C++ is going to make it much more verbose and more difficult to maintain. To test the thing properly, even the C++ test should check that the verification fails *for the right reason*, which means checking the error generated error message, so we're still pinning the error message. Not that I consider this a problem anyway. It's a very simple thing to change the error string embedded in the test if it's changed in the verification.

Reply:
For what concerns tests and pinning the error messages: I'm not saying that doing the test in C++ would set us free from the bounds of pinning error messages. And to be clear I think it's fine to pin error messages in this case.

What I was trying to say is:
1. I consider a superior goal to only have a single valid syntax for parsing, even more so if that syntax makes it impossible by design to write `switch` constructs with mismatched case values.
2. If we can get rid of the universal textual format for operations that we define custom parsers for, then that would make it impossible to test in `.mlir` files for `switch` operations with mismatched case values, because they wouldn't be expressible with a valid syntax
3. Even if we succeed in banning the universal parsing format for operations with custom serialization/deserialization, I still consider important to test for `switch` operations with mismatched cases, because those could be constructed in memory and not for parsing. For those, I think it's better to make the test in C++, because doing it in `.mlir` would force us to enable universal format parsing, which I'd like to avoid. I didn't say that doing the test in C++ was less verbose, or less difficult to maintain, or that it allowed not to pin error messages. I just meant that I consider writing the test in C++ (even if many others can be done in `.mlir` files) an acceptable price to pay if that's the only way to ban the universal parsing format, because the universal parsing format AFAIU isn't doing anything good for us that the custom parsing format can't already do.

---

Meanwhile, I've read some more about the universal MLIR format, form the link you pasted, and other material.
At the moment I really don't understand if there is any way to disable it, but I'd really like you to give a shot at removing it, only for `switch`es, or for operations with custom parsing.

> This is very much an escape hatch where the user is explicitly stepping outside of the guardrails.

I understand, but given that stepping outside the guardrails in this case doesn't do anything good (meaning that everything good you can already express with the custom parsing format), I'd really like users to fail to specify weird `switch`es with this format, and only use the allowed one, which already allows all the expressive power needed in order to express any valid constructs.

Reply:
I really think this is misguided to even try. There's even an `mlir-opt` flag to print all operations using the generic notation. If anything, the custom assembly format should probably be thought of as a secondary one for pretty printing.

> I'd really like you to give a shot at removing it

How hacky a solution are you going to be okay with? Let's say we can do it, at least in debug mode, by inspecting the call stack in some SwitchOp builder. Is that going too far?

Reply:
Ok I've investigated some more on this and I'm backing off on disabling the generic notation.

If there's a flag to force printing all operations with the generic notation, that's also a big red flag to forcing its removal. Basically every piece of the MLIR infrastructure expects that to exist and be a first-class citizen.

Reply:
At this point I have few questions on parsing and printing though.

So for `switch`es we've implemented custom serialization and deserialization. 

For printing:
1.How do we make sure we abide when the flag for forcing generic text format is passed? Is it just that the main MLIR printing logic select the "builtin" printer for the generic notation and not our custom print format?
2. If the printing of `case_values` for the generic notation is taken care of by MLIR itself, what does the last statement of `SwitchOp::print` (the one I started this thread from) do?

For parsing, I'll open another comment on the relevant lines.

Reply:
1. AFAIK, we don't need to do anything for the generic format, since it is generic. The MLIR library is just going to do its own thing and never involve our code.
2. That line prints an optional attribute dictionary containing all the attributes that were not printed manually as part of the custom assembly format, meaning all attributes except `case_values`. Remember that it is possible to attach arbitrary named attributes to operations.